### ISSUE TITLE
PDF p. 174, difficulty in first para. of "SO WHAT?"

### ISSUE THREAD

**Issue by wconrad** [robconery/feedback#101]
On PDF p. 174:

> You   might   be  thinking... so  what?   Who cares   if  you know    if  a   program will    ever    end
> (or not)?   The answer  is: a   lot of  people. You just    wrote   an  awesome matching
> algorithm   for a   dating  site.   Your    boss    wants   to  be  sure    it  will    always  return  a   match.
> There's no  way to  do  that.

This is a tough paragraph.  I think there's something wrong here, but I'm not sure, and I don't have a fix.

There are provably correct algorithms, so if you used such an algorithm, you should be able to tell your boss, "yes."  If we reword it to, "your boss wants _you to write a program_ to be sure it will always return a match," then the program could be made to recognize such algorithms _in some cases_ and be able to answer "yes."  But if a provably correct algorithm were not used, or if you or the program can't recognize it, then the answer would be "unknown."  

I _think_ it is correct that we can't write a program that, given an arbitrary matching algorithm, will never answer "unknown" when asked if the program will always return a match.  But my brain is twisted into a knot by this, so I'm not sure.

**Comment by robconery** [comment]
I rewrote it thus:

You might be thinking... _so what?_ Who cares if you know whether a program will end? The answer is: _a lot of people_. For example: you just wrote an awesome matching algorithm for a dating site. Your boss wants to be sure it will always return a match. _There's no way to do that_.

> There are provably correct algorithms, so if you used such an algorithm, you should be able to tell your boss, "yes."

Well... _no_ if I understand the halting problem correctly. You can't state for certain a program will always return... anything really. Will it crash... will it return more than 2? Undecidable.

> But my brain is twisted into a knot by this, so I'm not sure.

Nailed it :).

**Comment by wconrad** [comment]
Here's the argument going through my head.  This has nothing to do with the book, really...  I just want to write it down to see if it still makes sense:

Imagine if I create a null program--it does nothing.  It absolutely, positively will not halt.  I then teach my "will it halt?" program to recognize the null program, and report "yes" when when asked if the null program will halt.  I can also write a simple endless loop and teach my recognizer to report "no" when asked if that program will halt.  I can do this for any number of trivial, easy-to-recognize programs, such that my "will it halt" program can sometimes answer "yes" and sometimes "no."  For any program it does not recognize (which will probably be all real-world programs), it must of course answer "unknown."

The issue isn't that it's impossible for a program (or person) to _ever_ answer the halting problem, as I think I just showed.  The issue is that it isn't possible for the program to never answer "unknown" when presented an arbitrary program.  It will invariable sometimes (almost all the time, probably) have to answer "unknown."

That's what lead to my quibble: In some cases, you (the person) absolutely could tell your boss that the algorithm will return a match.  Use an algorithm that has been mathematically proven, and you can do that.  A recognizer that was looking for that algorithm could conceivably be able to answer "yes."  In practical terms, though, it's not doable--the recognizer would pretty much have to be coded to recognize the exact code of the match algorithm, or only trivial transformations of it.

**Comment by robconery** [comment]
I think the whole point of the halting problem is that (basically) there will always exist, in a theoretical sense, some input that can throw your routine into an undecided state.

So, in a real world sense - I guess I'll side with you on this one that you _could_ probably rely on a solid algorithm and some safeguards to make sure your matching returns something. You just can't _guarantee_ that because of the theoretical aspect.

In a theoretical sense (if I have this right), there does exist somewhere some input that will break that algorithm. This sits on the fact that a Turing Machine can calculate anything that can be calculated. This set includes your algorithm and the thing that will break it.

**Comment by robconery** [comment]
Keep going on this if you like - I'm going to close just for housekeeping :).

**Comment by wconrad** [comment]
Yes, that's pretty much what I'm saying.  I still don't think the words in the book about the matching algorithm quite match what you and I are saying here, but I don't think it's critical.  The more I think about it, the more I realize that my objection is, as I said earlier, a quibble--not really important.  Anyone who understood the section prior to the matching algorithm claim will understand what you mean here.
